<a href="https://colab.research.google.com/github/bhuvana-ak/huggingface_experiments/blob/main/Python_Coding_Arrays.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

DSA Coding Notebook

Python
Array basics:
**What Is an Array in Python?**
An array is also a data structure that stores a collection of items. Like lists, arrays are **ordered, mutable, enclosed in square brackets, and able to store non-unique items.**

But when it comes to the array's ability to store different data types, the answer is not as straightforward. It depends on the kind of array used.

To use arrays in Python, you need to import either an array module or a NumPy package.

1.   import array as arr
2.   import numpy as np


The Python array module requires all array elements to be of the same type. Moreover, to create an array, you'll need to specify a value type. In the code below, the "i" signifies that all elements in array_1 are integers:



```
# This is formatted as code
array_1 = arr.array("i", [3, 6, 9, 12])
print(array_1)
print(type(array_1))
```


array('i', [3, 6, 9, 12])
<class 'array.array'>

On the other hand, NumPy arrays support different data types. To create a NumPy array, you only need to specify the items (enclosed in square brackets, of course):
```
array_2 = np.array(["numbers", 3, 6, 9, 12])
print (array_2)
print(type(array_2))
```
['numbers' '3' '6' '9' '12']
<class 'numpy.ndarray'>
As you can see, array_2 contains one item of the string type (i.e., "numbers") and four integers.

In [7]:
!git config --global user.email "bhuvana.adurkannan@gmail.com"
!git config --global user.name "bhuvana-ak"

In [9]:
!echo "code_ws" >> README.md
!git init



Reinitialized existing Git repository in /content/.git/


In [ ]:
!git add README.md
!git commit -m "first commit"
!git branch -M main
!git remote add origin https://github.com/bhuvana-ak/code_ws.git
!git push -u origin main

In [6]:
!git push --set-upstream origin main
#!git pull origin main

fatal: could not read Username for 'https://github.com': No such device or address


In [ ]:
import array as arr

my_arr = arr.array("i",[0,1,2])
print(my_arr)

array('i', [0, 1, 2])


In [ ]:
my_arr.append(3)
print(my_arr)

array('i', [0, 1, 2, 3])


In [ ]:
my_arr.insert(4,4)
print(my_arr)

array('i', [0, 3, 4, 5, 4, 11, 11, 23])


In [ ]:
print(my_arr.index(11)) #returns first index for the value, throws error if element not found

5


In [ ]:
print(my_arr.count(11))

2


In [ ]:
my_arr.extend([11,23])
print(my_arr)

array('i', [0, 3, 4, 5, 11, 11, 23])


*pop takes index as input parameter*

In [ ]:
my_arr.pop(1)
print(my_arr)

array('i', [0, 2, 3, 4, 5])


*remove takes value as input parameter*

In [ ]:
my_arr.remove(11)
print(my_arr)

TypeError: array.remove() takes exactly one argument (0 given)

In [ ]:
del my_arr[6]
print(my_arr)

array('i', [0, 1, 2, 3, 4, 5])


# Coding questions

## Find the Repeated Number in an Array
Given an array with elements 1 to n and size n+1. containing 1 duplicate element. Find the duplicate element in the array without modifying the array and using only constant extra space

In [ ]:
import array as arr

my_arr = arr.array("i",[2,3,4,5,5])

*Brute force approach*

In [ ]:
for i in range(len(my_arr)):
  print("i",i,"my_arr[i]",my_arr[i])
  for j in range(i+1, len(my_arr)):
    print("i",i,"j",j,"my_arr[i]",my_arr[i],"my_arr[j]",my_arr[j])
    if my_arr[i] == my_arr[j]:
      print("Duplicate found",my_arr[i])
      break

i 0 my_arr[i] 2
i 0 j 1 my_arr[i] 2 my_arr[j] 3
i 0 j 2 my_arr[i] 2 my_arr[j] 4
i 0 j 3 my_arr[i] 2 my_arr[j] 5
i 1 my_arr[i] 3
i 1 j 2 my_arr[i] 3 my_arr[j] 4
i 1 j 3 my_arr[i] 3 my_arr[j] 5
i 2 my_arr[i] 4
i 2 j 3 my_arr[i] 4 my_arr[j] 5
i 3 my_arr[i] 5


Using a set to store duplicate.

**Time Complexity:**

The loop iterates n times, where n is the length of my_arr.
Inside the loop, the in operation for a set has an average time complexity of O(1).
Similarly, the add operation for a set also has an average time complexity of O(1).
**Therefore, the overall time complexity is dominated by the loop, resulting in O(n).**

**Space Complexity:**

The set my_set stores unique elements from my_arr. In the worst-case scenario, all elements in my_arr are unique.
Hence, the space used by my_set grows linearly with the input size.
**Therefore, the space complexity is O(n)**

In [ ]:
my_set = set()
for i in range(len(my_arr)):
  if my_arr[i] in my_set:
    print("Duplicate found",my_arr[i])
  else:
    my_set.add(my_arr[i])


Duplicate found 5


Approach with modifying array in place.
**Time Complexity:**

The loop iterates n times, where n is the length of my_arr.
All operations inside the loop (index calculation, comparison, assignment) take constant time.
**Therefore, the overall time complexity is dominated by the loop, resulting in O(n).**

**Space Complexity:**

The code modifies the input array my_arr in place.
It doesn't use any additional data structures that scale with the input size.
**Therefore, the space complexity is constant, denoted as O(1).**

In [ ]:
my_arr = arr.array("i",[1,2,3,4,5,6,6])
print(my_arr)
for i in range(len(my_arr)):
  index = abs(my_arr[i])-1
  print("i",i,"index",index,"my_arr[index]",my_arr[index])
  if my_arr[index] <0:
    print("Duplicate found",abs(my_arr[index]))
    break
  else:
    my_arr[index] = -my_arr[index]

array('i', [1, 2, 3, 4, 5, 6, 6])
i 0 index 0 my_arr[index] 1
i 1 index 1 my_arr[index] 2
i 2 index 2 my_arr[index] 3
i 3 index 3 my_arr[index] 4
i 4 index 4 my_arr[index] 5
i 5 index 5 my_arr[index] 6
i 6 index 5 my_arr[index] -6
Duplicate found 6


Let's assume the array contains value from 1 to n where each number is duplicated except 1.
nums = [1,2,3,2,1] find the one number that is not a duplicate.
output = 3

#Brute force


In [ ]:
nums = [1,1,2,2,3,3,4]
non_duplicate = -1
for i in range(len(nums)):
  is_duplicate = False
  for j in range(len(nums)):
    print("i",i,"j",j,"nums[i]",nums[i],"nums[j]",nums[j])
    if i !=j and nums[i] == nums[j]:
     is_duplicate = True
     break
  if not is_duplicate:
    print(nums[i])
    non_duplicate = nums[i]
    break

print("non_duplicate is ",non_duplicate)



i 0 j 0 nums[i] 1 nums[j] 1
i 0 j 1 nums[i] 1 nums[j] 1
i 1 j 0 nums[i] 1 nums[j] 1
i 2 j 0 nums[i] 2 nums[j] 1
i 2 j 1 nums[i] 2 nums[j] 1
i 2 j 2 nums[i] 2 nums[j] 2
i 2 j 3 nums[i] 2 nums[j] 2
i 3 j 0 nums[i] 2 nums[j] 1
i 3 j 1 nums[i] 2 nums[j] 1
i 3 j 2 nums[i] 2 nums[j] 2
i 4 j 0 nums[i] 3 nums[j] 1
i 4 j 1 nums[i] 3 nums[j] 1
i 4 j 2 nums[i] 3 nums[j] 2
i 4 j 3 nums[i] 3 nums[j] 2
i 4 j 4 nums[i] 3 nums[j] 3
i 4 j 5 nums[i] 3 nums[j] 3
i 5 j 0 nums[i] 3 nums[j] 1
i 5 j 1 nums[i] 3 nums[j] 1
i 5 j 2 nums[i] 3 nums[j] 2
i 5 j 3 nums[i] 3 nums[j] 2
i 5 j 4 nums[i] 3 nums[j] 3
i 6 j 0 nums[i] 4 nums[j] 1
i 6 j 1 nums[i] 4 nums[j] 1
i 6 j 2 nums[i] 4 nums[j] 2
i 6 j 3 nums[i] 4 nums[j] 2
i 6 j 4 nums[i] 4 nums[j] 3
i 6 j 5 nums[i] 4 nums[j] 3
i 6 j 6 nums[i] 4 nums[j] 4
4
non_duplicate is  4


Using set as a solution

In [ ]:
nums = [1,3,2,1,3,4,2]
my_set = set()
for i in range(len(nums)):
  if nums[i] in my_set:
    print("Duplicate found",nums[i])
    my_set.remove(nums[i])
  else:
    my_set.add(nums[i])

print("non_duplicate",my_set.pop())

Duplicate found 1
Duplicate found 3
Duplicate found 2
non_duplicate 4


math approach

In [ ]:
nums = [1,3,2,1,3,4,2]

n = (len(nums)+1)/2
sum = n *(n+1)
print(sum)

for i in range(len(nums)):
  sum -= nums[i]

print("non duplicate is : ",sum)

20.0
non duplicate is :  4.0


inplace array substitution if array is sorted

In [ ]:
nums = [1,1,2,2,3]
result = 0
for i in range(len(nums)):
  result ^= nums[i]

print("Non duplicate"result)

3


inplace array substitution is array is unsorted and n is known in advance

In [ ]:
nums = [1,2,3,2,1]
n =3
count = [0] *n
print(count)
for i in range(len(nums)):
  count[nums[i]-1] +=1

print(count)
for i in range(len(count)):
  if count[i] ==1:
    print("Non duplicate",i+1)

[0, 0, 0]
[2, 2, 1]
Non duplicate 3


Floyd's cycle-finding algorithm (also known as the "tortoise and hare" algorithm) **to work correctly in its standard form, the values in the array should represent indices within the same array, and thus range from 1 to n (or 0 to n-1, depending on indexing convention).**

Here's why:

**Pointer Movement:** *The algorithm relies on using the array values as pointers to traverse the array. Each value directs you to the next index to visit.*

**Cycle Detection:** *A cycle is detected when two pointers (the "tortoise" and the "hare") meet at the same index. This meeting point indicates a loop within the sequence of indices.*
If the values don't correspond to valid indices within the array, the pointers might go out of bounds or fail to detect cycles properly.

Phase 1: Finding the Meeting Point

This phase is about detecting the existence of a cycle. The fast pointer, moving twice as fast, is guaranteed to catch up to the slow pointer if there's a cycle.


Phase 2: Finding the Cycle's Entrance

Once we know there's a cycle, this phase pinpoints its starting point. The mathematical reasoning behind this phase is a bit involved, but the key idea is that by resetting one pointer and moving both at the same speed, they'll meet at the cycle's entrance.

In [ ]:
my_arr = arr.array("i",[1,3,4,2,4,5])
slow = fast = 0

#Cycle detection - Check if cycle exists

while True:
  slow = my_arr[slow]
  fast = my_arr[my_arr[fast]]
  print("slow",slow,"fast",fast)
  if slow== fast:
    print("Cycle detected", slow, fast)
    break

while slow !=fast:
  slow= my_arr[slow]
  fast= my_arr[fast]

print("Duplicate found",slow)



slow 1 fast 3
slow 3 fast 4
slow 2 fast 4
slow 4 fast 4
Cycle detected 4 4
Duplicate found 4


In [ ]:
nums = [1,2,2,3,4]
s = f = nums[0]
while True:
  s= nums[s]
  f=nums[nums[f]]
  if s==f:
    break

s = nums[0]
while s !=f:
  s= nums[s]
  f=nums[f]

print(s)

2


Given a string, find the length of the smallest window that contains every distinct character. Characters may appear more than once in the window.
For example, given "jiujitsu", you should return 5, corresponding to the final five letters.

In [ ]:
s ='abcccbab'
my_set = set(s)
print(my_set,len(my_set))



def min_window_brute_force(s):
    n = len(s)
    min_len = float('inf')

    for i in range(n):
        for j in range(i + 1, n+1):
            print("i",i,"j",j,"min_len",min_len)
            window = s[i:j]
            print("window",window)
            if len(set(window)) == len(set(s)):  # Check if all distinct chars are present
                min_len = min(min_len, len(window))

    return min_len if min_len != float('inf') else 0

{'a', 'b', 'c'} 3


In [ ]:
min_window_brute_force(s)

i 0 j 1 min_len inf
window a
i 0 j 2 min_len inf
window ab
i 0 j 3 min_len inf
window abc
i 0 j 4 min_len 3
window abcc
i 0 j 5 min_len 3
window abccc
i 0 j 6 min_len 3
window abcccb
i 0 j 7 min_len 3
window abcccba
i 0 j 8 min_len 3
window abcccbab
i 1 j 2 min_len 3
window b
i 1 j 3 min_len 3
window bc
i 1 j 4 min_len 3
window bcc
i 1 j 5 min_len 3
window bccc
i 1 j 6 min_len 3
window bcccb
i 1 j 7 min_len 3
window bcccba
i 1 j 8 min_len 3
window bcccbab
i 2 j 3 min_len 3
window c
i 2 j 4 min_len 3
window cc
i 2 j 5 min_len 3
window ccc
i 2 j 6 min_len 3
window cccb
i 2 j 7 min_len 3
window cccba
i 2 j 8 min_len 3
window cccbab
i 3 j 4 min_len 3
window c
i 3 j 5 min_len 3
window cc
i 3 j 6 min_len 3
window ccb
i 3 j 7 min_len 3
window ccba
i 3 j 8 min_len 3
window ccbab
i 4 j 5 min_len 3
window c
i 4 j 6 min_len 3
window cb
i 4 j 7 min_len 3
window cba
i 4 j 8 min_len 3
window cbab
i 5 j 6 min_len 3
window b
i 5 j 7 min_len 3
window ba
i 5 j 8 min_len 3
window bab
i 6 j 7 min_len 3
wi

3

In [ ]:
s ='abcbabb'
min_window_len = float('inf')

for i in range(len(s)):
  for j in range(i+1,len(s)+1):
    window = s[i:j]
    print("i",i,"j",j,"window",window)
    if len(set(window)) == len(set(s)):
      print("window",window)
      min_window_len = min(min_window_len,len(window))

print(min_window_len)


i 0 j 1 window a
i 0 j 2 window ab
i 0 j 3 window abc
window abc
i 0 j 4 window abcb
window abcb
i 0 j 5 window abcba
window abcba
i 0 j 6 window abcbab
window abcbab
i 0 j 7 window abcbabb
window abcbabb
i 1 j 2 window b
i 1 j 3 window bc
i 1 j 4 window bcb
i 1 j 5 window bcba
window bcba
i 1 j 6 window bcbab
window bcbab
i 1 j 7 window bcbabb
window bcbabb
i 2 j 3 window c
i 2 j 4 window cb
i 2 j 5 window cba
window cba
i 2 j 6 window cbab
window cbab
i 2 j 7 window cbabb
window cbabb
i 3 j 4 window b
i 3 j 5 window ba
i 3 j 6 window bab
i 3 j 7 window babb
i 4 j 5 window a
i 4 j 6 window ab
i 4 j 7 window abb
i 5 j 6 window b
i 5 j 7 window bb
i 6 j 7 window b
3


In [ ]:
def min_window_optimal(s):
    char_freq = {}
    window_start = 0
    min_window_len = float('inf')
    distinct_count = len(set(s)) # Count of distinct characters in the string

    for window_end in range(len(s)):
        right_char = s[window_end]
        char_freq[right_char] = char_freq.get(right_char, 0) + 1

        while len(char_freq) == distinct_count:
            min_window_len = min(min_window_len, window_end - window_start + 1)
            left_char = s[window_start]
            char_freq[left_char] -= 1
            if char_freq[left_char] == 0:
                del char_freq[left_char]
            window_start += 1

    return min_window_len if min_window_len != float('inf') else 0

In [ ]:
min_window_optimal(s)

3

In [ ]:
def min_window_optimal1(s):
   print(s)
   char_set = set()
   window_start = 0
   max_window_len = 0
   for window_end in range(len(s)):
    print("char_set",char_set)
    while(s[window_end] in char_set):
      char_set.remove(s[window_start])
      window_start +=1
    char_set.add(s[window_end])
    max_window_len = max(max_window_len, window_end-window_start+1)

   return max_window_len


In [ ]:
print(min_window_optimal1(s))

abcbabb
char_set set()
char_set {'a'}
char_set {'b', 'a'}
char_set {'b', 'a', 'c'}
char_set {'b', 'c'}
char_set {'b', 'a', 'c'}
char_set {'a', 'b'}
3


Move Zeros to End of Array
1. Use new array to save non zero array
T - O(n)
S - O(n)

2. sort the array and then split the array.
T- nlog(n)
S- O(1)
But no order is maintained

3. nested for loop o(n^2)


In [ ]:
nums = [0,1,0,3,2,0,4]
for i in range(len(nums)):
  for j in range(i+1,len(nums)):
    print("i",i,"j",j,"nums[i]",nums[i],"nums[j]",nums[j],"nums",nums)
    if nums[i] == 0 and nums[j] !=0:
      nums[i],nums[j] = nums[j],nums[i]
      break
print(nums)

i 0 j 1 nums[i] 0 nums[j] 1 nums [0, 1, 0, 3, 2, 0, 4]
i 1 j 2 nums[i] 0 nums[j] 0 nums [1, 0, 0, 3, 2, 0, 4]
i 1 j 3 nums[i] 0 nums[j] 3 nums [1, 0, 0, 3, 2, 0, 4]
i 2 j 3 nums[i] 0 nums[j] 0 nums [1, 3, 0, 0, 2, 0, 4]
i 2 j 4 nums[i] 0 nums[j] 2 nums [1, 3, 0, 0, 2, 0, 4]
i 3 j 4 nums[i] 0 nums[j] 0 nums [1, 3, 2, 0, 0, 0, 4]
i 3 j 5 nums[i] 0 nums[j] 0 nums [1, 3, 2, 0, 0, 0, 4]
i 3 j 6 nums[i] 0 nums[j] 4 nums [1, 3, 2, 0, 0, 0, 4]
i 4 j 5 nums[i] 0 nums[j] 0 nums [1, 3, 2, 4, 0, 0, 0]
i 4 j 6 nums[i] 0 nums[j] 0 nums [1, 3, 2, 4, 0, 0, 0]
i 5 j 6 nums[i] 0 nums[j] 0 nums [1, 3, 2, 4, 0, 0, 0]
[1, 3, 2, 4, 0, 0, 0]


Optimal O(n) solution
using 2 pointers - quick sort

In [ ]:
l=0
nums = [0,1,0,3,2,0,4,9,0,0,8]
for r in range(1,len(nums)):
  if nums[l]==0 and nums[r] !=0:
    nums[l],nums[r] = nums[r],nums[l]
  if nums[l] !=0:
    l+=1
print(nums)

[1, 3, 2, 4, 9, 8, 0, 0, 0, 0, 0]


In [ ]:
l=0
nums = [1,0,1,0,3,2,0,4,9,0,0,8]
for r in range(0,len(nums)):
  if nums[r]:
    print("l",l,"r",r,"nums[l]",nums[l],"nums[r]",nums[r],"nums",nums)
    nums[l],nums[r] = nums[r],nums[l]
    l+=1
print(nums)

l 0 r 0 nums[l] 1 nums[r] 1 nums [1, 0, 1, 0, 3, 2, 0, 4, 9, 0, 0, 8]
l 1 r 2 nums[l] 0 nums[r] 1 nums [1, 0, 1, 0, 3, 2, 0, 4, 9, 0, 0, 8]
l 2 r 4 nums[l] 0 nums[r] 3 nums [1, 1, 0, 0, 3, 2, 0, 4, 9, 0, 0, 8]
l 3 r 5 nums[l] 0 nums[r] 2 nums [1, 1, 3, 0, 0, 2, 0, 4, 9, 0, 0, 8]
l 4 r 7 nums[l] 0 nums[r] 4 nums [1, 1, 3, 2, 0, 0, 0, 4, 9, 0, 0, 8]
l 5 r 8 nums[l] 0 nums[r] 9 nums [1, 1, 3, 2, 4, 0, 0, 0, 9, 0, 0, 8]
l 6 r 11 nums[l] 0 nums[r] 8 nums [1, 1, 3, 2, 4, 9, 0, 0, 0, 0, 0, 8]
[1, 1, 3, 2, 4, 9, 8, 0, 0, 0, 0, 0]


My own question:
 Given an array / list of string with varying length, sort the list by len of string in descending order.

 1. brute force o(n^2) solution using nested for loops

> Add blockquote



In [ ]:
s = ["a","abc","ab","abcd",""]

for i in range(len(s)):
  for j in range(i+1,len(s)):
    if len(s[i]) < len(s[j]):
      s[i], s[j] = s[j], s[i]

print(s)

['abcd', 'abc', 'ab', 'a', '']


2 pointer solution

In [ ]:
s = ["a","abc","ab","abcd",""]
 s1 = []
l = 0
for r in range(len(s)):
  print("l",l,"r",r,"s[l]",s[l],"s[r]",s[r],"s",s)
  while len(s[l]) < len(s[r]):
    s[l],s[r] = s[r],s[l]
    l +=1
print(s)

l 0 r 0 s[l] a s[r] a s ['a', 'abc', 'ab', 'abcd', '']
l 0 r 1 s[l] a s[r] abc s ['a', 'abc', 'ab', 'abcd', '']
l 1 r 2 s[l] a s[r] ab s ['abc', 'a', 'ab', 'abcd', '']
l 2 r 3 s[l] a s[r] abcd s ['abc', 'ab', 'a', 'abcd', '']
l 3 r 4 s[l] a s[r]  s ['abc', 'ab', 'abcd', 'a', '']
['abc', 'ab', 'abcd', 'a', '']


prodct of all numbers except self

In [ ]:
nums=[1,2,3,4,5]

result = []
arr_len = len(nums)
pre_array = [1] * arr_len
post_array = [1] * arr_len

# pre_array = [1,1,1,1]
# post_array = [1,1,1,1]

for i in range(1, arr_len):
  pre_array[i] = pre_array[i-1] * nums[i-1]

  #pre_array = [1,2,6,24]

for i in range(arr_len-2,-1,-1):
  post_array[i] = post_array[i+1] * nums[i+1]

    #post_array = [60,20,5,1]
print(pre_array)
print(post_array)
for i in range(arr_len):
  result.append(pre_array[i] * post_array[i])

print(result)

[1, 1, 2, 6, 24]
[120, 60, 20, 5, 1]
[120, 60, 40, 30, 24]


using 1 array and 2 for loops

In [ ]:
nums=[2,3,4,5]
arr_len = len(nums)
result = [1] * arr_len

res =1
for i in range(arr_len):
  result[i] = res
  res *= nums[i]

res =1
for i in range(arr_len-1,-1,-1):
  result[i] *= res
  res *= nums[i]

print(result)

[60, 40, 30, 24]


3 sum in an array

In [ ]:
nums = [-1,2,3,-1,0,1,-2,-2]
arr_len = len(nums)

nums.sort()
print(nums)
target = 0
res= []
for i in range(arr_len):
  l,r= i+1, arr_len-1
  if i >0 and nums[i] == nums[i-1]:
    continue
  while l < r:
    a1,a2,a3=nums[i], nums[l],nums[r]
    three_sum = a1+a2+a3

    if three_sum ==0:
      res.append([i,l,r])
      l+=1
      if nums[l] == nums[l-1] and l < r :
        l+=1
       # break
    elif three_sum < target :
        l+=1
    elif three_sum > target:
        r-=1
print(res)


[-2, -2, -1, -1, 0, 1, 2, 3]
[[0, 2, 7], [0, 4, 6], [2, 3, 6], [2, 4, 5]]


Given an array nums consisting of only 0s, 1s, and 2s, sort the array in-place.

Input: nums = [2, 0, 2, 1, 1, 0]
Output: [0, 0, 1, 1, 2, 2]

In [ ]:
nums = [2, 0, 2, 1, 1, 0]
new_nums = [0]*len(nums)
zero_cntr = 0
for i in range(len(nums)):
  if nums[i] == 0:
    new_nums[zero_cntr] = 0
    zero_cntr += 1
print(zero_cntr)

one_cntr = zero_cntr
print("one_cntr",one_cntr)

for i in range(len(nums)):
  if nums[i] == 1:
    print("nums",nums[i])
    new_nums[one_cntr] = 1
    one_cntr +=1
print("one_cntr---",one_cntr)

two_cntr = one_cntr
for i in range(len(nums)):
  if nums[i] == 2:
    new_nums[two_cntr] = 2
    two_cntr +=1

print(two_cntr)
print(new_nums)

# BAD SOLUTION for in place array substitution

2
one_cntr 2
nums 1
nums 1
one_cntr--- 4
6
[0, 0, 1, 1, 2, 2]


In [ ]:
count_0 = count_1 = count_2 = 0
nums = [2,0,0,0,1,1,1,1,2,2]
for num in nums:
  if num == 0:
    count_0 += 1
  elif num == 1:
    count_1 += 1
  else:
    count_2 += 1

for i in range(count_0):
    nums[i] = 0
for i in range(count_0, count_0 + count_1):
    nums[i] = 1
for i in range(count_0 + count_1,len(nums)):
    nums[i] = 2

print(nums)

[0, 0, 0, 1, 1, 1, 1, 2, 2, 2]


In [ ]:
nums = [2, 0, 2, 1, 1, 0]
l,m,h=0,0,len(nums)-1
while m <h:
  if nums[m] == 0:
    nums[l] , nums[m] = nums[m], nums[l]
    m+=1
    l+=1
  elif nums[m] ==1:
    m+=1
  else:
    nums[h], nums[m] = nums[m], nums[h]
    h-=1

print(nums)

[0, 0, 1, 1, 2, 2]


Given an integer array nums and an integer k, return the k most frequent elements within the array.

The test cases are generated such that the answer is always unique.

You may return the output in any order.

Example 1:

Input: nums = [1,2,2,3,3,3], k = 2

Output: [2,3]

In [ ]:
hm = {}
nums = [1,2,2,3,3,3,3]
for i in range(len(nums)):
  a = nums[i]
  hm[a] = hm.get(a,0) +1
print(hm)
sorted_items = sorted(hm.items(), key=lambda item: item[1])
print(sorted_items.pop()[1])

{1: 1, 2: 2, 3: 4}
4


In [ ]:
filtered_items = dict(filter(lambda x : x[0]==1, hm.items()))
print(filtered_items)

{1: 1}


In [ ]:
new_hm = {key :x*x for key,x in hm.items()}
print(new_hm)

{1: 1, 2: 4, 3: 16}


In [ ]:
import heapq
k=2
nums = [1,2,2,3,3,3,3]
negated_data = [-x for x in nums]  # Negate values
heapq.heapify(negated_data)
arr_len = len(nums)
for i in range(k):
  heap= heapq.heappop(negated_data)
  print(heap)



-3
-3


Bucket sort

In [ ]:
nums = [1,1,2,2,3,3,3,3]
top_k = 2
counter = {}
freq = [[] for i in range(len(nums))]
print(freq)
for i in range(len(nums)):
  counter[nums[i]] = counter.get(nums[i],0)+1

print(counter)

for k,v in counter.items():
  print(k,v)
  freq[v].append(k)

print(freq)
res = []
for val in range(len(freq)-1,0,-1):
  if top_k ==0:
    break
  if freq[val]:
    res.append(freq[val])
    print(freq[val])
    top_k -=1

print(res)
print("freq",freq)
top_k = 2
res= []
for val in range(len(freq)-1,0,-1):
    print("val",val,"f[val]",freq[val])
    if freq[val]:
     res.append(freq[val])
    if len(res) == top_k:
      break

print("res",res)

[[], [], [], [], [], [], [], []]
{1: 2, 2: 2, 3: 4}
1 2
2 2
3 4
[[], [], [1, 2], [], [3], [], [], []]
[3]
[1, 2]
[[3], [1, 2]]
freq [[], [], [1, 2], [], [3], [], [], []]
val 7 f[val] []
val 6 f[val] []
val 5 f[val] []
val 4 f[val] [3]
val 3 f[val] []
val 2 f[val] [1, 2]
res [[3], [1, 2]]


Design an algorithm to encode a list of strings to a single string. The encoded string is then decoded back to the original list of strings.

Please implement encode and decode

In [ ]:
input = ["I","got","t#is!"]
def get_len(i):
  return str(len(i))
def encode():
  s = ""
  for i in input:
    s += get_len(i)+"#"+i
  return s
encode()


def decode(input):
  output = []
  i = 0
  while i < len(input):
    j =i
    while input[j] != '#':
      j+=1
    length = int(input[i:j])
    output.append(input[j+1:j+1+length])
    i = j+1+length
  print(output)
decode(encode())

['I', 'got', 't#is!']


Given an array of strings strs, group all anagrams together into sublists. You may return the output in any order.

An anagram is a string that contains the exact same characters as another string, but the order of the characters can be different.

Example 1:

Input: strs = ["act","pots","tops","cat","stop","hat"]

Output: [["hat"],["act", "cat"],["stop", "pots", "tops"]]

In [ ]:
strs = ["act","pots","tops","cat","stop","hat"]
ouput = []
for i in range(len(strs)):
  j = i+1
  res_i= {}
  for s in strs[i]:
    res_i[s] = res_i.get(s,0)+1
  print("res_i",res_i)
  while j < len(strs):
    res_j= {}
    if len(strs[i]) != len(strs[j]):
      j +=1
      continue
    else:
      for s in strs[j]:
        res_j[s] = res_j.get(s,0)+1
      print("res_j",res_j)
      if res_j == res_i:
        print("both dictionary are same")
        ouput.append([strs[i],strs[j]])
      j +=1
print(ouput)


res_i {'a': 1, 'c': 1, 't': 1}
res_j {'c': 1, 'a': 1, 't': 1}
both dictionary are same
res_j {'h': 1, 'a': 1, 't': 1}
res_i {'p': 1, 'o': 1, 't': 1, 's': 1}
res_j {'t': 1, 'o': 1, 'p': 1, 's': 1}
both dictionary are same
res_j {'s': 1, 't': 1, 'o': 1, 'p': 1}
both dictionary are same
res_i {'t': 1, 'o': 1, 'p': 1, 's': 1}
res_j {'s': 1, 't': 1, 'o': 1, 'p': 1}
both dictionary are same
res_i {'c': 1, 'a': 1, 't': 1}
res_j {'h': 1, 'a': 1, 't': 1}
res_i {'s': 1, 't': 1, 'o': 1, 'p': 1}
res_i {'h': 1, 'a': 1, 't': 1}
[['act', 'cat'], ['pots', 'tops'], ['pots', 'stop'], ['tops', 'stop']]


You are given a large web server log file where each line represents a single HTTP request. Each line contains fields such as timestamp, client IP address, requested URL, HTTP status code, and response time.

Design an algorithm to identify the top 10 most frequently requested URLs within a specific time window (e.g., the last hour). How would you handle the scenario where the log file is too large to fit into memory?

#Log file
2023-08-20 14:32:15 192.168.1.10 /index.html 200 120
2023-08-20 14:32:20 10.0.0.5 /products/item1.html 200 85
2023-08-20 14:32:35 192.168.1.10 /about.html 200 90
2023-08-20 14:32:40 172.16.0.20 /contact.html 404 50
2023-08-20 14:32:48 192.168.1.10 /index.html 200 115
2023-08-20 14:33:00 10.0.0.5 /products/item2.html 200 78


In [ ]:
# a function to process the file line by line and get the IP address and time (ensure mal formed lines are handled)
# process line only if time is within the 1 hour window
# as each Ip address is extracted add them to hashmap and increment the count
# The result from this is a hashmap with key =ip and value = count


# write another function that can iterate through this hashmap with keys and values.
#  create an list of size 10.
    # get the k,v from hashmap

    # check if there is a value in the index 0 if not add a tuple of (count ie v, ip ie k)
    # if value already exist, extract the tuple and get the count and compare against the value in hashmap.
        #if existing value is > hashmap value, replace value at given index with hashmap value and move the existing value in index to next position

# iterating through entire hashmap is not scalable solution as the number of lines increase with unique urls, adding these to hashmap and then iterating through hashmap wont be scalable

In [ ]:
import heapq
from datetime import datetime, timedelta

def top_n_urls(log_file_path, time_window_hours=1, n=1):
    """
    Identifies the top 10 most frequently requested URLs within a specific time window.

    Args:
        log_file_path: Path to the log file.
        time_window_hours: The time window in hours (default is 1 hour).

    Returns:
        A list of the top 10 URLs and their counts.
    """

    url_counts = {}  # Hashmap to store URL counts
    min_heap = []  # Min-heap to maintain the top 10 URLs
    url_to_heap_index = {}  # Map to track URL positions in the heap
    end_time =datetime(2023,8,20,15,00,15) #datetime.now()
    start_time = end_time - timedelta(hours=time_window_hours)

    with open(log_file_path, 'r') as log_file:
        for line in log_file:
            try:
                timestamp_dt, timestamp_time, ip,url, _ = line.split(maxsplit=4)

                timestamp = datetime.strptime(timestamp_dt+" "+timestamp_time, '%Y-%m-%d %H:%M:%S')
                if start_time <= timestamp <= end_time:
                    url_counts[url] = url_counts.get(url, 0) + 1
                    print("min heap len? - n",len(min_heap),n,url)
                    if url in url_to_heap_index:
                        print("url in url_to_heap_index",url,url_counts[url])
                        # URL already in heap, update its count and reposition in heap
                        index = url_to_heap_index[url]
                        print("min_heap[index]",index,min_heap[index])
                        min_heap[index] = (url_counts[url], url)
                        #heapq._siftdown(min_heap, 0, index)  # Restore heap property
                        heapq.heapify(min_heap)
                        print(heapq.nlargest(n, min_heap))
                    elif len(min_heap) < n:
                        # Heap not full, add new URL
                        heapq.heappush(min_heap, (url_counts[url], url))
                        url_to_heap_index[url] = len(min_heap) - 1
                    elif url_counts[url] > min_heap[0][0]:
                        # Replace root if new URL has higher count
                        _, old_url = heapq.heapreplace(min_heap, (url_counts[url], url))
                        del url_to_heap_index[old_url]
                        url_to_heap_index[url] = 0

            except ValueError:
                # Handle malformed lines gracefully (e.g., log or skip)
                pass
            except Exception as e:
                print("Error while file processing",e)
    print("url_counts: ",url_counts,min_heap)
    print("min_heap",min_heap)
    top_n_urls = [(url, count) for count, url in heapq.nlargest(n, min_heap)]
    return top_n_urls

# Example usage:
log_file_path = 'my_log_file.txt'
top_urls = top_n_urls(log_file_path,n=2)
print("top_urls-",top_urls)

min heap len? - n 0 2 /index.html
min heap len? - n 1 2 /products/item1.html
min heap len? - n 2 2 /about.html
min heap len? - n 2 2 /contact.html
min heap len? - n 2 2 /index.html
url in url_to_heap_index /index.html 2
min_heap[index] 0 (1, '/index.html')
[(2, '/index.html'), (1, '/products/item1.html')]
min heap len? - n 2 2 /products/item2.html
min heap len? - n 2 2 /index.html
url in url_to_heap_index /index.html 3
min_heap[index] 0 (1, '/products/item1.html')
[(3, '/index.html'), (2, '/index.html')]
min heap len? - n 2 2 /products/item1.html
url in url_to_heap_index /products/item1.html 2
min_heap[index] 1 (3, '/index.html')
[(2, '/products/item1.html'), (2, '/index.html')]
min heap len? - n 2 2 /about.html
min heap len? - n 2 2 /contact.html
min heap len? - n 2 2 /index.html
url in url_to_heap_index /index.html 4
min_heap[index] 0 (2, '/index.html')
[(4, '/index.html'), (2, '/products/item1.html')]
min heap len? - n 2 2 /products/item2.html
url_counts:  {'/index.html': 4, '/produ

You are given a list of tasks where each task is represented by a tuple (task_id, priority, duration).  The task_id is a unique identifier, priority is an integer representing the task's importance (higher value indicates higher priority), and duration is an integer representing the time required to complete the task.

Implement a function schedule_tasks(tasks) that returns the order in which the tasks should be executed to minimize the total waiting time. The total waiting time is calculated as the sum of the time each task waits before its execution starts. Assume that only one task can be executed at a time.

Example:

Input: tasks = [(1, 3, 5), (2, 1, 2), (3, 2, 3)]
Output: [1, 3, 2]

Explanation:
- Task 1 has the highest priority, so it's executed first. Total waiting time = 0
- Task 3 has the next highest priority. Total waiting time = 5 (it waits for task 1 to complete)
- Task 2 is executed last. Total waiting time = 5 + 3 = 8

The total waiting time for this schedule is 0 + 5 + 8 = 13
Hints:

You can use a priority queue (min-heap or max-heap depending on your implementation) to efficiently keep track of the tasks based on their priority or remaining duration.
Consider how you will handle ties in priority.
Additional Considerations (for more advanced questions):

What if tasks arrive dynamically over time?
How would you handle preemption (interrupting a running task to execute a higher priority task)?
This question effectively tests a candidate's understanding of priority queues, their ability to design algorithms, and their problem-solving skills.

Feel free to ask if you would like a more specific solution or another coding question involving priority queues!

In [ ]:
import heapq

def schedule_tasks(tasks):
    task_queue = [(-priority, task_id, duration) for task_id, priority, duration in tasks]
    heapq.heapify(task_queue)

    current_time = 0
    execution_order = []
    completed_tasks = {}  # To keep track of completed tasks and their completion times

    while task_queue:
        priority, task_id, duration = heapq.heappop(task_queue)

        execution_order.append(task_id)
        current_time += duration
        completed_tasks[task_id] = current_time

        # Update remaining durations only for tasks not yet completed
        task_queue = [(p, tid, max(0, d - (current_time - completed_tasks.get(tid, 0))))
                       for p, tid, d in task_queue if tid not in completed_tasks]
        #task_queue = [(p, tid, d)
                      #for p, tid, d in task_queue if tid not in completed_tasks]
        heapq.heapify(task_queue)

    return execution_order

In [ ]:
schedule_tasks( [(1, 3, 5), (2, 1, 2), (3, 2, 8)] )

[1, 3, 2]

Task Scheduling
You are given an array of CPU tasks tasks, where tasks[i] is an uppercase english character from A to Z. You are also given an integer n.

Each CPU cycle allows the completion of a single task, and tasks may be completed in any order.

The only constraint is that identical tasks must be separated by at least n CPU cycles, to cooldown the CPU.

Return the minimum number of CPU cycles required to complete all tasks.

Example 1:

Input: tasks = ["X","X","Y","Y"], n = 2

Output: 5
Explanation: A possible sequence is: X -> Y -> idle -> X -> Y.

Example 2:

Input: tasks = ["A","A","A","B","C"], n = 3

Output: 9
Explanation: A possible sequence is: A -> B -> C -> Idle -> A -> Idle -> Idle -> Idle -> A.

Constraints:

1 <= tasks.length <= 1000
0 <= n <= 100


In [ ]:
from collections import deque, Counter
from heapq import heapify, heappush, heappop
tasks = ["A","A","A","B","C","B","C"]
n = 3

cntr = Counter(tasks)
print(cntr)
max_heap = [-c for c in cntr.values()]
heapify(max_heap)
print(max_heap)
time = 0
q = deque()
while max_heap or q:

  if max_heap: # if there are values in max heap
    task_count = heappop(max_heap) +1 # represents count of each task
    time += 1 #Time for each task
    if task_count: # add to queue only if task count is more than 0 (frequency of task)
      q.append([task_count,time+n])

  else :
    time = q[0][1]


  if q and q[0][1]==time:
    tsk_cnt = q.popleft()[0]
    heappush(max_heap,tsk_cnt)


print("time",time)

# while maxHeap or q:
#             time += 1

#             if not maxHeap:
#                 time = q[0][1]
#             else:
#                 cnt = 1 + heapq.heappop(maxHeap)
#                 if cnt:
#                     q.append([cnt, time + n])
#             if q and q[0][1] == time:
#                 heapq.heappush(maxHeap, q.popleft()[0])
#         return time



Counter({'A': 3, 'B': 2, 'C': 2})
[-3, -2, -2]
time 9
